[gdbm] Add new port, starting on version 1.24

noahknegt opened the pull request:
This PR adds a new port for the GDBM package, I need this as a required port as avahi depends on it which I would like to add after this.

- [x] Changes comply with the [maintainer guide](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/contributing/maintainer-guide.md).
- [x] The name of the port matches an existing name for this component on https://repology.org/ if possible, and/or is strongly associated with that component on search engines.
- [x] Optional dependencies are resolved in exactly one way. For example, if the component is built with CMake, all `find_package` calls are REQUIRED, are satisfied by `vcpkg.json`'s declared dependencies, or disabled with [CMAKE_DISABLE_FIND_PACKAGE_Xxx](https://cmake.org/cmake/help/latest/variable/CMAKE_DISABLE_FIND_PACKAGE_PackageName.html).
- [x] The versioning scheme in `vcpkg.json` matches what upstream says.
- [x] The license declaration in `vcpkg.json` matches what upstream says.
- [x] The installed as the "copyright" file matches what upstream says.
- [x] The source code of the component installed comes from an authoritative source.
- [x] The generated "usage text" is accurate. See [adding-usage](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/examples/adding-usage.md) for context.
- [x] The version database is fixed by rerunning `./vcpkg x-add-version --all` and committing the result.
- [x] Only one version is in the new port's versions file.
- [x] Only one version is added to each modified port's versions file.

noahknegt committed:
Add gdbm port

Signed-off-by: Noah Knegt <git@noahknegt.com>

noahknegt committed:
Add gdbm to version database

Signed-off-by: Noah Knegt <git@noahknegt.com>

dg0yt commented on ports/gdbm/portfile.cmake:
```diff
@@ -0,0 +1,55 @@
+vcpkg_download_distfile(ARCHIVE
+    URLS "https://ftp.gnu.org/gnu/gdbm/gdbm-${VERSION}.tar.gz"
```
```suggestion
    URLS "https://ftp.gnu.org/gnu/gdbm/gdbm-${VERSION}.tar.gz"
         "https://www.mirrorservice.org/sites/ftp.gnu.org/gnu/gdbm/gdbm-${VERSION}.tar.gz"
```

noahknegt replied:
Added multiple mirrors to the url list.

dg0yt commented on ports/gdbm/portfile.cmake:
```diff
@@ -0,0 +1,55 @@
+
+vcpkg_extract_source_archive(
+    SOURCE_PATH
+    ARCHIVE ${ARCHIVE}
```
```suggestion
    ARCHIVE "${ARCHIVE}"
```

noahknegt replied:
Applied feedback

dg0yt commented on ports/gdbm/portfile.cmake:
```diff
@@ -0,0 +1,55 @@
+    ARCHIVE ${ARCHIVE}
+)
+
+vcpkg_check_features(OUT_FEATURE_OPTIONS FEATURE_OPTIONS
+  FEATURES
+    libgdbm-compat      ENABLE_GDBM_COMPAT
+    readline            ENABLE_READLINE
+    memory-mapped-io    ENABLE_MMAP
+)
+
```
```suggestion
```

noahknegt replied:
Applied feedback

dg0yt commented on ports/gdbm/portfile.cmake:
```diff
@@ -0,0 +1,55 @@
+    "--enable-gdbmtool-debug=no"
+)
+
+if(${ENABLE_GDBM_COMPAT})
```
```suggestion
if("libgdbm-compat" IN_LIST FEATURES)
```

noahknegt replied:
Applied feedback

dg0yt commented on ports/gdbm/portfile.cmake:
```diff
@@ -0,0 +1,55 @@
+    list(APPEND options "--enable-libgdbm-compat=yes")
+endif()
+
+if(NOT ${ENABLE_READLINE})
```
```suggestion
if("readline" IN_LIST FEATURES)
    list(APPEND options "--with-readline")
else()
```

noahknegt replied:
Applied feedback, with some slight modifications. Readline is automatically enabled when found on the system, and thus to avoid using it implicitly the condition is reversed. Also added a comment with a link to the original docs that explains this.

dg0yt commented on ports/gdbm/portfile.cmake:
```diff
@@ -0,0 +1,55 @@
+    list(APPEND options "--without-readline")
+endif()
+
+if(NOT ${ENABLE_MMAP})
```
```suggestion
if("memory-mapped-io" IN_LIST FEATURES)
    list(APPEND options "--enable-memory-mapped-io")
else()
```

noahknegt replied:
Applied feedback, with some slight modifications. Mmap is automatically enabled when found on the system, and thus to avoid using it implicitly the condition is reversed. Also added a comment with a link to the original docs that explains this.

dg0yt commented on ports/gdbm/vcpkg.json:
```diff
@@ -0,0 +1,25 @@
+      "description": "Enable the use of mmap(2) for I/O optimizations."
+    },
+    "readline": {
+      "description": "Enable GNU Readline support."
```
```suggestion
      "description": "Enable GNU Readline support.",
      "dependencies": [
        "readline"
      ]
```

noahknegt replied:
Applied feedback

dg0yt reviewed (commented):
Community feedback. Not tested.

noahknegt committed:
Apply feedback provided in a review

This feedback includes the following:
- *FTP mirrors are added*
- *Readline is added as a dependency*
- *Archive variable is put in quotes to prevent word splitting*

Co-authored-by: @dg0yt
Signed-off-by: Noah Knegt <git@noahknegt.com>

noahknegt committed:
rework how the features are checked

Signed-off-by: Noah Knegt <git@noahknegt.com>

noahknegt commented:
> @noahknegt Have you tested all the features? I encountered an error while testing the feature: `(error: Autoconf version 2.71 or higher is required)`, but the `autoconf` version I can install to is `2.69`.

I have tested all the features, but done that on my WSL running `ubuntu 24.04`. It seems that I have `autoconf version 2.71` so that why it probably worked when I was testing it. I've seen no specific requirements for a autoconf version in the [readme](https://git.gnu.org.ua/gdbm.git/tree/?h=v1.24) of  `gdbm` so I would not know why it does not work. 

Is there any way the port can enforce a minimum version of `autoconf` through the `portfile.cmake`, @Mengna-Li ?

noahknegt committed:
Update version of the git-tree

Signed-off-by: Noah Knegt <git@noahknegt.com>

noahknegt commented:
> Community feedback. Not tested.

Thanks @dg0yt for the feedback most of it I applied and worked, made some small changes to the suggestions as for both `readline` and `memory-mapped-io` these are enabled by default, see [readme](https://git.gnu.org.ua/gdbm.git/tree/README?h=v1.24), when detected on the system. So to avoid using them implicitly when they are found I chose to disable them when the feature is not present.

dg0yt commented:
"Detected when present" may need explicit control also to ensure failure when the feature is requested from the port but configuration fails to detect it.

> Is there any way the port can enforce a minimum version of autoconf through the portfile.cmake

AFAICT no such check is present in any other port. The log should be clear enough in case of error.

Ubuntu 20.04 comes with autoconf 2.69. I wonder why @Mengna-Li even tests with this old version. And it is possible to use a local installation of a newer version.

noahknegt commented:
> "Detected when present" may need explicit control also to ensure failure when the feature is requested from the port but configuration fails to detect it.

How would you do this properly, not that familiar with creating vcpkg ports yet nor with `autoconf`

dg0yt commented:
> How would you do this properly, not that familiar with creating vcpkg ports yet nor with `autoconf

There is no general answer without checkout the configure scripts. autoconf build systems vary a lot. The basic assumptions is that you need to control both cases. if/else for enable/disable, with/without. That's what I suggested for readline.

noahknegt committed:
Add more feedback from the PR

Signed-off-by: Noah Knegt <git@noahknegt.com>

noahknegt committed:
Updated git-tree for gdbm

Signed-off-by: Noah Knegt <git@noahknegt.com>

noahknegt commented:
> > How would you do this properly, not that familiar with creating vcpkg ports yet nor with `autoconf
> 
> There is no general answer without checkout the configure scripts. autoconf build systems vary a lot. The basic assumptions is that you need to control both cases. if/else for enable/disable, with/without. That's what I suggested for readline.

I seem to have missed the else in your suggested changes, but have now applied it. Just verified in the [configure.ac](https://git.gnu.org.ua/gdbm.git/tree/configure.ac?h=v1.24#n144) that `autoconf` will produce errors when the features are requested but not present on the system. 

So right now all of your feedback has been applied and tested @dg0yt, thanks again :)
